drivers/dp/swdp_bitbang.c

```diff
@@ -0,0 +1,632 @@
+/*
+ * Copyright (c) 2018-2019 PHYTEC Messtechnik GmbH
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This file is based on SW_DP.c from DAPLink Interface Firmware.
+ * Copyright (c) 2009-2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+/* Serial Wire Debug Port interface bit-bang driver */
+
+#define DT_DRV_COMPAT zephyr_swdp_gpio
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/swdp.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(swdp, CONFIG_DP_DRIVER_LOG_LEVEL);
+
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+#define CPU_CLOCK		64000000U
+#else
+#define CPU_CLOCK		CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#endif
+
+#define CLOCK_DELAY(swclk_freq, port_write_cycles) \
+	((CPU_CLOCK / 2 / swclk_freq) - port_write_cycles)
+
+/*
+ * Default SWCLK frequency in Hz.
+ * sw_clock can be used to overwrite this default value.
+ */
+#define SWDP_DEFAULT_SWCLK_FREQUENCY	1000000U
+
+#define DELAY_SLOW_CYCLES		3U
+
+struct sw_config {
+	struct gpio_dt_spec clk;
+	struct gpio_dt_spec dout;
+	struct gpio_dt_spec din;
+	struct gpio_dt_spec dnoe;
+	struct gpio_dt_spec noe;
+	struct gpio_dt_spec reset;
+	uint32_t port_write_cycles;
+	void *clk_reg;
+	void *dout_reg;
+	void *din_reg;
+	void *dnoe_reg;
+};
+
+struct sw_cfg_data {
+	uint32_t clock_delay;
+	uint8_t turnaround;
+	bool data_phase;
+	bool fast_clock;
+};
+
+static uint8_t sw_request_lut[16] = {0U};
+
+static void mk_sw_request_lut(void)
+{
+	uint32_t parity = 0U;
+
+	for (int request = 0; request < sizeof(sw_request_lut); request++) {
+		parity = request;
+		parity ^= parity >> 2;
+		parity ^= parity >> 1;
+
+		/*
+		 * Move A[3:3], RnW, APnDP bits to their position,
+		 * add start bit, stop bit(6), and park bit.
+		 */
+		sw_request_lut[request] =  BIT(7) | (request << 1) | BIT(0);
+		/* Add parity bit */
+		if (parity & 0x01U) {
+			sw_request_lut[request] |= BIT(5);
+		}
+	}
+
+	LOG_HEXDUMP_DBG(sw_request_lut, sizeof(sw_request_lut), "request lut");
+}
+
+static uint32_t ALWAYS_INLINE sw_get32bit_parity(uint32_t data)
+{
+	data ^= data >> 16;
+	data ^= data >> 8;
+	data ^= data >> 4;
+	data ^= data >> 2;
+	data ^= data >> 1;
+
+	return data & 1U;
+}
+
+static ALWAYS_INLINE void pin_delay_asm(uint32_t delay)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+	__asm volatile ("movs r3, %[p]\n"
+			".start_%=:\n"
+			"subs r3, #1\n"
+			"bne .start_%=\n"
+			:
+			: [p] "r" (delay)
+			: "r3", "cc"
+			);
+#else
+#error "Not defined for this SoC family"
+#endif
+}
+
+static ALWAYS_INLINE void pin_platform_set(void *base, uint8_t pin)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+	NRF_GPIO_Type * reg = base;
+
+	reg->OUTSET = BIT(pin);
+#else
+#error "Not defined for this SoC family"
+#endif
+}
```

Comment:
I'm a bit confused by these platform specific functions. Does every SoC we want to support need to be added here? Can't we abstract this using existing drivers + devicetree?

Reply:
Would assume it is for speed, you need these GPIO transitions to be fast, using the zephyr drivers would be universal but probably add far too much overhead/delay. There is a description of this in the DAPLINK repository.

Reply:
Yes, it is what @nordicjm says. I made it more generic as in my previous PR and throughput has worsened. But bit-bang driver is rather for initial work and verification, we have at least two platforms where much faster implementation using specific peripherals is possible.

Reply:
Would be cool to have a sample on the RP2040 using the PIO system.